bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/internal/AudioServlet.java

```diff
@@ -195,13 +228,42 @@ public String serve(AudioStream stream) {
     }
 
     @Override
-    public String serve(FixedLengthAudioStream stream, int seconds) {
+    public String serve(AudioStream stream, int seconds) {
         String streamId = UUID.randomUUID().toString();
-        multiTimeStreams.put(streamId, stream);
+        if (stream instanceof FixedLengthAudioStream fixedLengthAudioStream) {
+            multiTimeStreams.put(streamId, fixedLengthAudioStream);
+        } else { // we prefer a FixedLengthAudioStream, but we can try to make one
+            try {
+                multiTimeStreams.put(streamId, createFixedLengthInputStream(stream, streamId));
+            } catch (AudioException | IOException e) {
+                logger.warn("Cannot precache the audio stream to serve it", e);
+            }
+        }
         streamTimeouts.put(streamId, System.nanoTime() + TimeUnit.SECONDS.toNanos(seconds));
         return getRelativeURL(streamId);
     }
 
+    private FixedLengthAudioStream createFixedLengthInputStream(AudioStream stream, String streamId)
+            throws IOException, AudioException {
+        byte[] dataBytes = stream.readNBytes(ONETIME_STREAM_BUFFER_MAX_SIZE + 1);
+        FixedLengthAudioStream fixedLengthAudioStreamResult;
+        if (dataBytes.length <= ONETIME_STREAM_BUFFER_MAX_SIZE) {
+            // we will use an in memory buffer to avoid disk operation
+            fixedLengthAudioStreamResult = new ByteArrayAudioStream(dataBytes, stream.getFormat());
+        } else {
+            // in memory max size exceeded, sound is too long, we will use a file
+            File tempFile = File.createTempFile(streamId, ".snd");
+            tempFile.deleteOnExit();
+            try (OutputStream outputStream = new FileOutputStream(tempFile)) {
+                outputStream.write(dataBytes);
```

Comment:
This will only write the first ONETIME_STREAM_BUFFER_MAX_SIZE + 1 bytes and not all ?

Reply:
To avoid using too much memory, I indeed set a limit. But only for the in-memory buffer.

If the limit is reached, then the 1MB memory buffer is copied to a file and the remaining of the stream is also copied to it.

Reply:
Ok, this is done by the next line (stream.transferTo)